components/scheduler2/core/scheduler.go

```diff
@@ -0,0 +1,171 @@
+)
+
+var (
+	_ hsm.StateMachine[enumsspb.Scheduler2State] = Scheduler{}
+	_ hsm.StateMachineDefinition                 = &machineDefinition{}
+)
+
+// Registers state machine definitions with the HSM registry. Should be called
+// during dependency injection.
+func RegisterStateMachines(r *hsm.Registry) error {
+	if err := r.RegisterMachine(machineDefinition{}); err != nil {
+		return err
+	}
+	// TODO: add other state machines here
+	return nil
+}
+
+// MachineCollection creates a new typed [statemachines.Collection] for operations.
+func MachineCollection(tree *hsm.Node) hsm.Collection[Scheduler] {
+	return hsm.NewCollection[Scheduler](tree, SchedulerMachineType)
+}
+
+func (s Scheduler) State() enumsspb.Scheduler2State {
+	return s.HsmSchedulerV2State.State
+}
+
+func (s Scheduler) SetState(state enumsspb.Scheduler2State) {
+	s.HsmSchedulerV2State.State = state
+}
+
+func (s Scheduler) RegenerateTasks(node *hsm.Node) ([]hsm.Task, error) {
+	return nil, nil
+}
+
+func (machineDefinition) Type() string {
+	return SchedulerMachineType
+}
+
+func (machineDefinition) Serialize(state any) ([]byte, error) {
+	if state, ok := state.(Scheduler); ok {
+		return proto.Marshal(state.HsmSchedulerV2State)
+	}
+	return nil, fmt.Errorf("invalid scheduler state provided: %v", state)
+}
+
+func (machineDefinition) Deserialize(body []byte) (any, error) {
+	state := &schedspb.HsmSchedulerV2State{}
+	return Scheduler{
+		HsmSchedulerV2State: state,
+		compiledSpec:        nil,
+	}, proto.Unmarshal(body, state)
+}
+
+// Returns:
+//
+// 0 when states are equal
+// 1 when a is newer than b
+// -1 when b is newer than a
+func (machineDefinition) CompareState(a any, b any) (int, error) {
+	s1, ok := a.(Scheduler)
+	if !ok {
+		return 0, fmt.Errorf("%w: expected state1 to be a Scheduler instance, got %v", hsm.ErrIncompatibleType, s1)
+	}
+	s2, ok := a.(Scheduler)
+	if !ok {
+		return 0, fmt.Errorf("%w: expected state1 to be a Scheduler instance, got %v", hsm.ErrIncompatibleType, s2)
+	}
+
+	if s1.State() > s2.State() {
+		return 1, nil
+	} else if s1.State() < s2.State() {
+		return -1, nil
+	}
+
+	return 0, nil
+}
+
+// Returns true when the Scheduler should allow scheduled actions to be taken.
+//
+// When decrement is true, the schedule's state's `RemainingActions` counter
+// is decremented and the conflict token is bumped.
+func (s Scheduler) CanTakeScheduledAction(decrement bool) bool {
+	// If paused, don't do anything
+	if s.Schedule.State.Paused {
+		return false
+	}
+
+	// If unlimited actions, allow
+	if !s.Schedule.State.LimitedActions {
+		return true
+	}
+
+	// Otherwise check and decrement limit
+	if s.Schedule.State.RemainingActions > 0 {
+		if decrement {
+			s.Schedule.State.RemainingActions--
+			s.ConflictToken++
+		}
+		return true
+	}
+
+	// No actions left
+	return false
+}
+
+func (s Scheduler) CompiledSpec(specBuilder *scheduler.SpecBuilder) (*scheduler.CompiledSpec, error) {
+	// cache compiled spec
+	if s.compiledSpec == nil {
```

Comment:
Use the conflict token number to know if your cache may be invalid instead of worrying about manual invalidation?

Reply:
This isn't for manual invalidation, it's to support lazily loading the compiled spec after deserializing the state machine. Ideally, I'd fill this in at `Deserialize` time, but that would require I have access to the `SpecBuilder` (which itself caches time zone files from disk).. I suppose I could wire in `SpecBuilder` onto the `machineDefinition`, but I think I remember you mentioning that CHASM automates the `Deserialize`/`Serialize` bits away from components, right?

Reply:
What if the spec is updated later, don't you need to recompile it?

> CHASM automates the Deserialize/Serialize bits away from components

Yes, which means you're right about not wanting to put this in deserialize, but also for the reason I gave above.

Reply:
> What if the spec is updated later, don't you need to recompile it?

Yeah, you do; I see your point, will update.